Add viewport override to 2D editor

KoBeWi opened the pull request:
This PR supersedes #61873 with alternative approach.
Instead of enabling other viewports and fighting with the editor to make it work properly, I added a button that allows to override the default 2D editor viewport with another one.

https://user-images.githubusercontent.com/2223172/223558255-e3a3c033-6b92-4f7d-b116-c56c1124a6a5.mp4

I have one unresolved problem though - the overriding viewport's size must match the editor's size to display properly. Right now I'm just displaying its content with a TextureRect, but maybe I should use some different approach, idk.
I didn't make much progress for some time, so I'm opening a draft.

Closes https://github.com/godotengine/godot-proposals/issues/2139
Closes https://github.com/godotengine/godot-proposals/issues/5422
Closes #39387

KoBeWi committed:
Add viewport override to 2D editor

KoBeWi commented:
Ok this is somewhat finished. I fixed the viewport problems by adding more viewports. So the 2D editor structure is now SubViewportContainer -> SubViewport -> TextureRect -> current viewport.

The current viewport is still displayed inside TextureRect, but screen transform is now different. While using main viewport, the displayed viewports' transform is changed. If the viewport is overriden, the containing viewport (the one under SubViewportContainer) is transformed instead and the overriding viewport is displayed in local space. It works quite well.

https://user-images.githubusercontent.com/2223172/233101084-95d37ee7-1784-4421-80f4-07d409d5992d.mp4

I'm really surprised that selecting nodes works without any problems in this setup.

The only problem are Window nodes. For whatever reason they don't redraw when content changes and don't draw anything while invisible.

https://user-images.githubusercontent.com/2223172/233101280-5c6e959b-1824-4fc5-a0da-530471e6c47b.mp4

(I also remembered that the override button is buggy)

But it's ready to be ready for review I think.

jordanlis commented:
This is amazing, really need that one :)

Variable-ind commented:
Indeed It'll Help A LOT

Summersay415 commented:
What's status of this?

Summersay415 commented:
Tested locally, works good with some caveats you mentioned. Suggestion: how about not restricting rendering outside of the overridden viewport, as in a normal, not overridden one?

KoBeWi commented:
This is a technical restriction. The main viewport works a bit differently - when you move around, its canvas is transformed and displays a different region. The rendering is still restricted to the viewport's size, but it has the size of the editor, so you don't see that objects outside don't render.

jan-Kulpas commented:
Would this also work for 3D nodes with a 2D Subviewport?

KoBeWi commented:
No, this doesn't cover 3D.

AndreaTerenz commented:
Looks really good! 
But what about multiple subviewports in the same scene? Would it be a simple extension to turn the toggle button into a dropdown to select a specific subview?

KoBeWi commented:
Currently you have to select SubViewport node and click the button on the toolbar. You can preview any SubViewport in the scene using that.

Calinou reviewed (commented):
Tested locally, it works as expected in all rendering methods.

It even works with viewports that contain 3D nodes, as long as it's previewed within the 2D editor :slightly_smiling_face: 

![image](https://github.com/user-attachments/assets/4a4b9859-f07c-471f-bc9d-7ae09b48da33)

Some feedback:

- It's hard to distinguish between the button being disabled (when a non-Viewport node is selected), enabled, and active:

Disabled | Enabled | Active
-|-|-
![image](https://github.com/user-attachments/assets/cea177f0-9d80-4783-98ad-2aa8eae3736e) | ![image](https://github.com/user-attachments/assets/0a92dd32-7f95-4e4c-9de3-bbd4c04ea7b0) |  ![image](https://github.com/user-attachments/assets/8141daad-a1c5-4f3f-8693-1d8ff16eada0)

I suggest using a copy of the icon that is white, so that it's more visible (and there's higher contrast between the disabled and enabled state).

- There should be tooltip when hovering the aforementioned button. Something like this:

```
Viewport Override
Overrides the 2D editor view with the contents of the selected Viewport node.
```

This would match the tooltip of the Project Camera Override button nicely (which is right besides it).

KoBeWi commented:
The button was using old style. Changed it to use the new FlatButton:

https://github.com/user-attachments/assets/1de3e585-c73f-45b8-89e2-d5430d966303

I also added tooltips:
![image](https://github.com/user-attachments/assets/f8242c3a-7a2a-4acb-9c65-195e1bd0361c)
![image](https://github.com/user-attachments/assets/defc45d2-eb5a-47e4-87d2-8309d0eeae15)
![image](https://github.com/user-attachments/assets/2128de74-dfe3-4124-9140-eb50bd3e5b5c)

I also found a bug where main screen plugin does not receive `edit(null)`, causing invalid button state when node is deselected. It should be fixed separately.

kitbdev commented:
- also fixes https://github.com/godotengine/godot/issues/39387
- related, but 3D https://github.com/godotengine/godot-proposals/issues/1465

KoBeWi commented:
Rebased, but the root viewport is bugged for some reason. Its ViewportTexture is smaller than the viewport, so the 2D editor view is cropped.

KoBeWi commented:
Rebased and integrated with #99401

AThousandShips commented on editor/scene/canvas_item_editor_plugin.cpp:
```diff
@@ -5574,11 +5700,19 @@ CanvasItemEditor::CanvasItemEditor() {
 	scene_tree->set_stretch(true);
 	scene_tree->set_anchors_and_offsets_preset(Control::PRESET_FULL_RECT);
-	scene_tree->add_child(EditorNode::get_singleton()->get_scene_root());
+	scene_tree->connect("item_rect_changed", callable_mp(this, &CanvasItemEditor::_update_viewport_size));
```
```suggestion
	scene_tree->connect(SceneStringName(item_rect_changed), callable_mp(this, &CanvasItemEditor::_update_viewport_size));
```

AThousandShips replied:
```suggestion
	scene_tree->connect(SceneStringName(item_rect_changed), callable_mp(this, &CanvasItemEditor::_update_viewport_size));
```

AThousandShips commented on editor/scene/canvas_item_editor_plugin.cpp:
```diff
@@ -5938,6 +6072,16 @@ CanvasItemEditor::CanvasItemEditor() {
 	main_menu_hbox->add_child(memnew(VSeparator));
 
+	override_viewport_button = memnew(Button);
+	override_viewport_button->set_theme_type_variation("FlatButton");
```
```suggestion
	override_viewport_button->set_theme_type_variation(SceneStringName(FlatButton));
```

Calinou reviewed (commented):
Tested locally, it works as expected.

Some feedback:

- The viewport icon is hard to distinguish from the background when active, due to its 1-pixel thick outline:

<img width="208" height="60" alt="Image" src="https://github.com/user-attachments/assets/ec5fc4df-b07f-408e-9b0e-28b9dc07e103" />

The icon is also darker than other icons besides it. Swapping this icon for an icon with a 2-pixel thick outline would improve things significantly here.

You can use this separate icon for the toolbar, with the same brightness as other icons (`ViewportOverride.svg`): <img width="16" height="16" alt="ViewportOverride" src="https://github.com/user-attachments/assets/b9fc9aa3-0df0-4e4b-a010-e5f686cf4d65" />

- When active, the tooltip could display the name of the Viewport node that is being used as an override. Since you can select different nodes without switching active viewports, it won't always match your selection.

  - For example:

    > Editor viewport is overridden to show "MySubviewport". Press to disable override.

KoBeWi commented:
I added a dedicated Scene dock button and also fixed some bugs.

<img width="966" height="142" alt="image" src="https://github.com/user-attachments/assets/c9b7b438-e494-41fc-a981-d13245f0ba32" />

> Some feedback:

Addressed.

adamscott commented on editor/scene/scene_tree_editor.cpp:
```diff
@@ -712,6 +715,10 @@ void SceneTreeEditor::_update_node(Node *p_node, TreeItem *p_item, bool p_part_o
 			p_item->add_button(0, get_editor_theme_icon(SNAME("Group")), BUTTON_GROUP, false, TTR("Children are not selectable.\nClick to make them selectable."));
 		}
 
+		if (_is_overriding_viewport(p_node)) {
+			p_item->add_button(0, get_editor_theme_icon(SNAME("ViewportOverride")), BUTTON_VIEWPORT, false, TTR("This viewport overrides the editor's viewport.\nClick to disable override."));
+		}
```
```suggestion
		Viewport *viewport_node = Object::cast_to<Viewport>(p_node);
		if (viewport_node != nullptr) {
			if (_is_overriding_viewport(p_node)) {
				p_item->add_button(0, get_editor_theme_icon(SNAME("ViewportOverrideEnabled")), BUTTON_VIEWPORT, false, TTR("This viewport overrides the editor's viewport.\nClick to disable override."));
			} else {
				p_item->add_button(0, get_editor_theme_icon(SNAME("ViewportOverrideDisabled")), BUTTON_VIEWPORT, false, TTR("This viewport doesn't override the editor's viewport.\nClick to enable override."));
			}
			_update_viewport_override_color(p_node, p_item);
		}
```

with this added code:
```cpp
void SceneTreeEditor::_update_viewport_override_color(Node *p_node, TreeItem *p_item) {
	Viewport *node_viewport = Object::cast_to<Viewport>(p_node);
	if (node_viewport == nullptr) {
		return;
	}

	Color color(1, 1, 1, 1);
	if (!_is_overriding_viewport(p_node)) {
		color.a = 0.6;
	}
	int idx = p_item->get_button_by_id(0, BUTTON_VIEWPORT);
	p_item->set_button_color(0, idx, color);
}

```

adamscott commented on editor/icons/ViewportOverride.svg:
I would rename this file `ViewportOverrideEnabled.svg` and I would create `ViewportOverrideDisabled.svg` with something like the following:

<img width="100%" alt="Image" src="https://github.com/user-attachments/assets/bcd95d19-eca7-4031-876d-d24a4edeee51" />

adamscott commented on editor/scene/scene_tree_editor.cpp:
```diff
@@ -322,6 +323,8 @@ void SceneTreeEditor::_cell_button_pressed(Object *p_item, int p_column, int p_i
 		} else {
 			_revoke_unique_name();
 		}
+	} else if (p_id == BUTTON_VIEWPORT) {
+		CanvasItemEditor::get_singleton()->set_viewport_override(nullptr);
 	}
```
```suggestion
	} else if (p_id == BUTTON_VIEWPORT) {
		if (_is_overriding_viewport(n)) {
			CanvasItemEditor::get_singleton()->set_viewport_override(nullptr);
		} else {
			CanvasItemEditor::get_singleton()->set_viewport_override(Object::cast_to<Viewport>(n));
		}
	}
```

adamscott commented on editor/scene/canvas_item_editor_plugin.h:
```diff
@@ -550,8 +568,12 @@ class CanvasItemEditor : public VBoxContainer {
+	void _button_override_viewport(bool p_pressed);
 	void _button_tool_select(int p_index);
 
+	void _update_override_viewport_button(Viewport *p_viewport);
```
```suggestion
```

adamscott commented on editor/scene/canvas_item_editor_plugin.h:
```diff
@@ -550,8 +568,12 @@ class CanvasItemEditor : public VBoxContainer {
 	void _button_toggle_local_space(bool p_status);
 	void _button_toggle_smart_snap(bool p_status);
 	void _button_toggle_grid_snap(bool p_status);
+	void _button_override_viewport(bool p_pressed);
```
```suggestion
```

adamscott commented on editor/scene/canvas_item_editor_plugin.h:
```diff
@@ -361,6 +371,7 @@ class CanvasItemEditor : public VBoxContainer {
 	Button *group_button = nullptr;
 	Button *ungroup_button = nullptr;
 
+	Button *override_viewport_button = nullptr;
```
```suggestion
```

adamscott commented on editor/scene/canvas_item_editor_plugin.cpp:
```diff
@@ -3060,6 +3159,11 @@ void CanvasItemEditor::_update_lock_and_group_button() {
 
+void CanvasItemEditor::set_viewport_override(Viewport *p_viewport) {
+	_set_viewport_override(p_viewport);
+	_update_override_viewport_button(edited_viewport);
```
```suggestion
```

adamscott commented on editor/scene/canvas_item_editor_plugin.cpp:
```diff
@@ -4397,6 +4506,10 @@ void CanvasItemEditor::_notification(int p_what) {
 			pivot_button->set_tooltip_text(TTR("Click to change object's pivot.") + "\n" + TTR("Shift: Set temporary pivot.") + "\n" + TTR("Click this button while holding Shift to put the temporary pivot in the center of the selected nodes.") + "\n" + show_list_tooltip);
 			pan_button->set_tooltip_text(TTR("You can also use Pan View shortcut (Space by default) to pan in any mode.") + "\n" + show_list_tooltip);
 			ruler_button->set_tooltip_text(TTR("LMB+Drag: Measure the distance between two points in 2D space.") + "\n" + show_list_tooltip);
+
+			if (_is_viewport_overridden()) {
+				_update_override_viewport_button(edited_viewport);
+			}
```
```suggestion
```

adamscott commented on editor/scene/canvas_item_editor_plugin.cpp:
```diff
@@ -4562,6 +4678,11 @@ void CanvasItemEditor::edit(CanvasItem *p_canvas_item) {
 
+void CanvasItemEditor::edit_viewport(Viewport *p_viewport) {
+	edited_viewport = p_viewport;
+	_update_override_viewport_button(p_viewport);
```
```suggestion
```

adamscott commented on editor/scene/canvas_item_editor_plugin.cpp:
```diff
@@ -4698,6 +4819,15 @@ void CanvasItemEditor::_button_toggle_grid_snap(bool p_status) {
+	} else {
+		_set_viewport_override(nullptr);
+	}
+	_update_override_viewport_button(edited_viewport);
```
```suggestion
```

adamscott commented on editor/scene/canvas_item_editor_plugin.cpp:
```diff
@@ -5508,6 +5673,11 @@ void CanvasItemEditor::set_state(const Dictionary &p_state) {
+
+	// Turn off viewport override when changing scenes etc.
+	_set_viewport_override(nullptr);
+	_update_override_viewport_button(nullptr);
```
```suggestion
```

adamscott commented on editor/scene/canvas_item_editor_plugin.cpp:
```diff
@@ -4818,6 +4948,41 @@ void CanvasItemEditor::_prepare_view_menu() {
 	popup->set_item_disabled(popup->get_item_index(CLEAR_GUIDES), !has_guides);
 }
 
+void CanvasItemEditor::_update_override_viewport_button(Viewport *p_viewport) {
+	const String override_message = vformat(TTR("Editor Viewport Override\nEditor viewport is overridden by \"%s\". Press to disable override."), current_viewport->get_name());
+	if (_is_viewport_valid(p_viewport)) {
+		override_viewport_button->set_disabled(false);
+		if (p_viewport == current_viewport) {
+			override_viewport_button->set_pressed_no_signal(true);
+			override_viewport_button->set_tooltip_text(override_message);
+		} else {
+			override_viewport_button->set_pressed_no_signal(false);
+			override_viewport_button->set_tooltip_text(TTRC("Editor Viewport Override\nMakes editor use the selected Viewport instead of the main viewport, allowing to edit nodes inside it."));
+		}
+	} else {
+		if (_is_viewport_overridden()) {
+			override_viewport_button->set_disabled(false);
+			override_viewport_button->set_pressed_no_signal(true);
+			override_viewport_button->set_tooltip_text(override_message);
+		} else {
+			override_viewport_button->set_disabled(true);
+			override_viewport_button->set_pressed_no_signal(false);
+			if (p_viewport) {
+				override_viewport_button->set_tooltip_text(TTRC("Editor Viewport Override\nThe selected Viewport is not suitable for overriding.\nSelect a SubViewport or a visible Window to allow overriding editor viewport."));
+			} else {
+				override_viewport_button->set_tooltip_text(TTRC("Editor Viewport Override\nNo Viewport node selected. Select one to allow overriding editor viewport."));
+			}
+		}
+	}
+}
+
```
```suggestion
```

adamscott commented on editor/scene/canvas_item_editor_plugin.cpp:
```diff
@@ -5953,6 +6133,16 @@ CanvasItemEditor::CanvasItemEditor() {
 
 	main_menu_hbox->add_child(memnew(VSeparator));
 
+	override_viewport_button = memnew(Button);
+	override_viewport_button->set_theme_type_variation(SceneStringName(FlatButton));
+	override_viewport_button->set_toggle_mode(true);
+	override_viewport_button->set_disabled(true);
+	main_menu_hbox->add_child(override_viewport_button);
+	override_viewport_button->connect(SceneStringName(toggled), callable_mp(this, &CanvasItemEditor::_button_override_viewport));
+	_update_override_viewport_button(nullptr);
+
```
```suggestion
```

adamscott reviewed (commented):
I would instead remove the icon from the toolbar and put it always besides a viewport. It would show that it's toggleable right away.


https://github.com/user-attachments/assets/b82d4603-7e8f-43b2-b6da-a40cc365405d

KoBeWi commented:
> I would instead remove the icon from the toolbar and put it always besides a viewport. It would show that it's toggleable right away.

The problem with that is that override only has effect for 2D. Not sure if 3D override even makes sense, as viewports share 3D worlds by default.
